dockers/docker-fpm-frr/bgpcfgd

```diff
@@ -1,65 +1,125 @@
 #!/usr/bin/env python
 
 import sys
+import copy
+import Queue
 import redis
 import subprocess
 import syslog
-from swsssdk import ConfigDBConnector
+import os
+from swsscommon import swsscommon
 
-class BGPConfigDaemon:
+
+def run_command(command):
+    syslog.syslog(syslog.LOG_DEBUG, "execute command {}.".format(command))
+    p = subprocess.Popen(command, shell=True, stdout=subprocess.PIPE)
+    stdout = p.communicate()[0]
+    p.wait()
+    if p.returncode != 0:
+        syslog.syslog(syslog.LOG_ERR, 'command execution returned {}. Command: "{}", stdout: "{}"'.format(p.returncode, command, stdout))
+
+
+class BGPConfigManager(object):
+    def __init__(self, daemon):
+        self.daemon = daemon
+        self.bgp_asn = None
+        self.bgp_message = Queue.Queue(0)
+        daemon.add_manager(swsscommon.CONFIG_DB, swsscommon.CFG_DEVICE_METADATA_TABLE_NAME, self.__metadata_handler)
+        daemon.add_manager(swsscommon.CONFIG_DB, swsscommon.CFG_BGP_NEIGHBOR_TABLE_NAME, self.__bgp_handler)
+
+    def __metadata_handler(self, key, op, data):
+        if key != "localhost" \
+            or "bgp_asn" not in data \
+                or self.bgp_asn == data["bgp_asn"]:
+            return
+        syslog.syslog(syslog.LOG_INFO, 'ASN changed to {} from {}, restart BGP...'.format(data['bgp_asn'], self.bgp_asn))
+
+        # TODO add ASN update commands 
+
+        self.bgp_asn = data["bgp_asn"]
+        self.__update_bgp()
+
+    def __update_bgp(self):
+        while not self.bgp_message.empty():
+            key, op, data = self.bgp_message.get()
+            syslog.syslog(syslog.LOG_INFO, 'value for {} changed to {}'.format(key, data))
+            if op == swsscommon.SET_COMMAND:
+                command = "vtysh -c 'configure terminal' -c 'router bgp {}' -c 'neighbor {} remote-as {}'".format(self.bgp_asn, key, data['asn'])
+                run_command(command)
+                if "name" in data:
+                    command = "vtysh -c 'configure terminal' -c 'router bgp {}' -c 'neighbor {} description {}'".format(self.bgp_asn, key, data['name'])
+                    run_command(command)
```

Comment:
What if this doesn't succeed?

Reply:
In the run_command function, we have printed out warning if command failed.

Reply:
The warning does nothing to handle the failure case. This is inherently problematic because it allows for DB to be out of sync with running config of the routing application. The config event will be lost and the only way to recover this (since it's not saved in startup config either) is to restart bgpcfgd. In general this code doesn't seem to be robust and the requirements are not well defined/clear.